File: jekyun-park/week5/programmers/86053.swift
//
//  main.swift
//  SwiftAlgorithms
//
//  Created by 박제균 on 2022/01/31.
//

import Foundation

// (도시에있는) g 금무게 s 은무게 t 편도이동시간 w 최대운반무게
// a 필요한 금의양, b 필요한 은의양
func solution(_ a: Int, _ b: Int, _ g: [Int], _ s: [Int], _ w: [Int], _ t: [Int]) -> Int64 {

    var start: Int = 0
    var end: Int = Int(10e15)
    var answer: Int = end

    while start <= end {
        // mid 는 이동시간을 나타냄
        let mid = (start + end) / 2
        var gold = 0
        var silver = 0
        var total = 0

        for i in 0..<s.count {
            // count: 이동시간 내에 운반할 수 있는 횟수
            var count = mid / (t[i] * 2)

            if mid % (t[i] * 2) >= t[i] {
                count += 1
            }
            
            // 시간내에 운반가능한 무게와 각 트럭의 금,은의 무게를 비교
            gold += min(g[i], w[i] * count)
            silver += min(s[i], w[i] * count)
            total += min(g[i] + s[i], w[i] * count)
        }
        
        // 위에서 구한 모든 트럭의 광물의 합의 최대치가 조건을 만족하는지 검사
        // a,b : 필요한 금,은의 양
        // a
        if gold >= a && silver >= b && total >= a + b {
            end = mid - 1
            answer = min(mid, answer)
        } else {
            start = mid + 1
        }
        
    }
    return Int64(answer)
}

print(solution(10, 10, [100], [100], [7], [10]))
print(solution(90, 500, [70, 70, 0], [0, 0, 500], [100, 100, 2], [4, 8, 1]))
// 10    10    [100]    [100]    [7]    [10]
// 90    500    [70,70,0]    [0,0,500]    [100,100,2]    [4,8,1]

File: leehyowonzero/5week/86053.py
def solution(a, b, g, s, w, t):
    start = 0
    end = 10**15
    answer = end

    while start <= end :
        mid = (start + end) // 2
        now_gold = 0
        now_silver = 0
        now_total = 0

        for i in range(len(t)):
            movabletimes = (mid - t[i]) // (t[i] * 2) + 1 # 첫 운반 고려 + 나머지는 왕복이 이루어져야 1번 카운트, 예외상황은 발생하지 않을것으로 판단
            
            # 금으로 다 옮길 때, 최대한 옮길 수 있는 만큼 옮기기
            if (movabletimes * w[i] > g[i]):  # 광물이 부족하니 바닥까지 긁어모아
                now_gold += g[i]
            else: # 광물이 남아...
                now_gold += movabletimes * w[i]
            # 은으로 다 옮길 때, 최대한 옮길 수 있는 만큼 옮기기    
            if (movabletimes * w[i] > s[i]):
                now_silver += s[i]
            else:
                now_silver += movabletimes * w[i]
            # 금 은 상관없이 옮길때, 최대한 옮길 수 있는 만큼 옮기기
            if (s[i] + g[i] < movabletimes * w[i]):
                now_total += s[i] + g[i]
            else:
                now_total += movabletimes * w[i]
                
        if (now_total >= a + b and now_gold >= a and now_silver >= b): # 정해진 시간동안 넉넉하게 광물을 옮길 수 있음
            end = mid - 1 # 시간단축
            answer = min(answer, mid)
        else:
            start = mid + 1 # 시간부족이니 증가

    return answer